File: projects/01-voice-receptionist/integrazioni/notifiche.js
#!/usr/bin/env node
/**
 * Adapter notifiche — WhatsApp/SMS al titolare via Twilio Messaging API.
 *
 * Finché le credenziali in .env sono "MOCK_..." (default), questo modulo
 * SIMULA l'invio: logga cosa avrebbe mandato, senza fare chiamate di rete
 * né spendere credito. Con credenziali vere, la stessa funzione fa una
 * vera richiesta HTTPS all'API Twilio — zero dipendenze npm, solo il
 * modulo https nativo di Node.
 */
'use strict';
const https = require('https');
const { CONFIG, modalitaSimulata } = require('./config');

function postForm(host, pathname, auth, campi) {
  return new Promise((resolve, reject) => {
    const corpo = new URLSearchParams(campi).toString();
    const req = https.request(
      {
        host,
        path: pathname,
        method: 'POST',
        auth, // 'accountSid:authToken' — Basic Auth richiesta da Twilio
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Content-Length': Buffer.byteLength(corpo),
        },
      },
      (res) => {
        let data = '';
        res.on('data', (c) => (data += c));
        res.on('end', () => resolve({ status: res.statusCode, body: data }));
      }
    );
    req.on('error', reject);
    req.write(corpo);
    req.end();
  });
}

/**
 * Invia una notifica WhatsApp al titolare. Ritorna { simulato: bool, ... }
 * così chi chiama sa sempre se è successo davvero o solo loggato.
 */
async function inviaWhatsAppTitolare(messaggio) {
  const { twilio, notifiche } = CONFIG;

  if (modalitaSimulata('twilio')) {
    console.log(`📵 [SIMULATO — credenziali mock] WhatsApp al titolare (${notifiche.numeroTitolare}):\n   "${messaggio}"`);
    return { simulato: true, messaggio };
  }

  const risposta = await postForm(
    'api.twilio.com',
    `/2010-04-01/Accounts/${twilio.accountSid}/Messages.json`,
    `${twilio.accountSid}:${twilio.authToken}`,
    {
      From: twilio.numeroWhatsapp,
      To: `whatsapp:${notifiche.numeroTitolare}`,
      Body: messaggio,
    }
  );
  return { simulato: false, status: risposta.status, body: risposta.body };
}

/** Compone il messaggio di notifica a partire da un evento del motore. */
function messaggioPerEvento(evento) {
  const dati = evento.dati || {};
  if (evento.tipo === 'appuntamento_creato') {
    return `📅 Nuovo appuntamento: ${dati.nome || 'cliente'} — preferenza: ${dati.preferenza_orario || 'da confermare'}. Gestito dall'assistente AI.`;
  }
  if (evento.tipo === 'urgenza_notificata') {
    return `🚨 URGENZA segnalata da ${dati.nome || 'chiamante sconosciuto'} — richiamare appena possibile.`;
  }
  if (evento.tipo === 'richiesta_stato_veicolo') {
    return `🔧 Richiesta stato veicolo, targa ${evento.targa || 'n/d'} — verificare e rispondere al cliente.`;
  }
  return `ℹ️ Evento "${evento.tipo}" gestito dall'assistente AI.`;
}

module.exports = { inviaWhatsAppTitolare, messaggioPerEvento };

File: projects/04-micro-saas-officine/integrazioni/notifiche.py
#!/usr/bin/env python3
"""Adapter SMS — avvisa il cliente via Twilio quando il veicolo è pronto.

Stesso pattern del progetto 01 (notifiche.js): finché le credenziali in
.env sono ancora "MOCK_..." (default), l'invio è SIMULATO (loggato in
console, nessuna chiamata di rete, nessun credito speso). Con credenziali
Twilio vere, la stessa funzione fa una vera richiesta HTTPS — zero
dipendenze esterne, solo urllib della libreria standard.

Uso tipico (da app.py, quando uno stato passa a "pronto"):
    from integrazioni.notifiche import invia_sms_pronto
    invia_sms_pronto(intervento)
"""
import base64
import json
import os
import urllib.request
import urllib.parse
import urllib.error


def _config():
    return {
        "account_sid": os.environ.get("TWILIO_ACCOUNT_SID", "MOCK_ACxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"),
        "auth_token": os.environ.get("TWILIO_AUTH_TOKEN", "MOCK_authtoken"),
        "numero_da": os.environ.get("TWILIO_NUMERO_SMS", "MOCK_+15551234567"),
    }


def _e_mock(cfg):
    return any(str(v).startswith("MOCK_") for v in cfg.values())


def _invia_sms(numero_a, testo):
    """Ritorna sempre {"simulato": bool, ...} così chi chiama sa se è
    successo davvero o solo stato loggato."""
    cfg = _config()
    if _e_mock(cfg):
        print(f"📵 [SIMULATO — credenziali Twilio mock] SMS a {numero_a}:\n   \"{testo}\"")
        return {"simulato": True, "testo": testo}

    corpo = urllib.parse.urlencode({"From": cfg["numero_da"], "To": numero_a, "Body": testo}).encode()
    auth = base64.b64encode(f"{cfg['account_sid']}:{cfg['auth_token']}".encode()).decode()
    req = urllib.request.Request(
        f"https://api.twilio.com/2010-04-01/Accounts/{cfg['account_sid']}/Messages.json",
        data=corpo,
        headers={"Authorization": f"Basic {auth}", "Content-Type": "application/x-www-form-urlencoded"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=10) as res:
            return {"simulato": False, "status": res.status, "body": res.read().decode()}
    except urllib.error.HTTPError as e:
        return {"simulato": False, "status": e.code, "body": e.read().decode()}


def invia_sms_pronto(intervento):
    """intervento: dict con almeno targa, modello, telefono (del cliente)."""
    testo = (
        f"Ciao! Il tuo veicolo {intervento.get('modello', '')} (targa {intervento.get('targa', '')}) "
        f"è pronto per il ritiro. Ti aspettiamo in officina."
    )
    return _invia_sms(intervento.get("telefono"), testo)


def invia_sms_richiamo(veicolo, tipo_intervento, giorni_da_ultimo):
    """Usato da richiami.py per il promemoria di tagliando/revisione."""
    testo = (
        f"Ciao! Sono passati {giorni_da_ultimo} giorni dall'ultimo {tipo_intervento} "
        f"del tuo veicolo (targa {veicolo.get('targa', '')}). Vuoi prenotare un controllo?"
    )
    return _invia_sms(veicolo.get("telefono"), testo)


if __name__ == "__main__":
    print(json.dumps(invia_sms_pronto({
        "targa": "AB123CD", "modello": "Panda 1.2", "telefono": "+393331234567",
    }), ensure_ascii=False, indent=2))
